jest-rs/src/main/java/io/searchbox/rs/client/http/JestRsClient.java

```diff
@@ -0,0 +1,153 @@
+     */
+    protected <T extends JestResult> Invocation prepareRequest(final Action<T> clientRequest) {
+        String elasticSearchRestUrl = getRequestURL(getNextServer(), clientRequest.getURI());
+
+        log.finest(MessageFormat.format("Request method={0} url={1}", clientRequest.getRestMethodName(), elasticSearchRestUrl));
+
+        Builder requestBuilder = httpClient.target(elasticSearchRestUrl).request(MediaType.APPLICATION_JSON);
+        // add headers added to action
+        for (Entry<String, Object> header : clientRequest.getHeaders().entrySet()) {
+            requestBuilder.header(header.getKey(), header.getValue().toString());
+        }
+
+        String data = clientRequest.getData(gson);
+        if (data == null || data.isEmpty()) {
+            return requestBuilder.build(clientRequest.getRestMethodName());
+        } else {
+            if ("GET".equals(clientRequest.getRestMethodName())) {
+                return requestBuilder.buildPost(Entity.entity(data, MediaType.APPLICATION_JSON_TYPE));
+            } else {
+                return requestBuilder.build(clientRequest.getRestMethodName(), Entity.entity(data, MediaType.APPLICATION_JSON_TYPE));
+            }
+        }
+    }
+
+    private <T extends JestResult> T deserializeResponse(Response response, Action<T> clientRequest) throws IOException {
+        return clientRequest.createNewElasticSearchResult(
+                response.getEntity() == null ? null : response.readEntity(String.class),
+                response.getStatus(),
+                response.getStatusInfo().getReasonPhrase(),
+                gson
+        );
+    }
+
+    public Client getHttpClient() {
+        return httpClient;
+    }
+
+    public void setHttpClient(Client httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    public Gson getGson() {
+        return gson;
+    }
+
+    public void setGson(Gson gson) {
+        this.gson = gson;
+    }
+
+    protected class DefaultCallback<T extends JestResult> implements InvocationCallback<Response> {
+        private final Action<T> clientRequest;
+        private final JestResultHandler<? super T> resultHandler;
+
+        public DefaultCallback(Action<T> clientRequest, JestResultHandler<? super T> resultHandler) {
+            this.clientRequest = clientRequest;
+            this.resultHandler = resultHandler;
+        }
+
+        @Override
+        public void completed(final Response response) {
+            T jestResult = null;
+            try {
+                jestResult = deserializeResponse(response, clientRequest);
+            } catch (IOException e) {
+                failed(e);
+            }
+            if (jestResult != null) resultHandler.completed(jestResult);
+        }
+
+        @Override
+        public void failed(final Throwable ex) {
+            log.log(Level.SEVERE, "Exception occurred during async execution.", ex);
+            // JEST-COMMON should really accept throwable not just exception
```

Comment:
If you would like to discuss it please open an issue, having this comment within the code serves no useful purpose.